src/ncoutputwriter.jl

```diff
@@ -0,0 +1,150 @@
+"""
```

Comment:
Is this a constructor for a struct called `WriteGeometry`? I think it is not, in which case it should use `snake_case`. 

In the rest of the code, I think we use the word `grid` to refer to the information that is saved by this function. We haven't introduced the concept of `geometry` --- I'm not sure this is the right place. What do you think?

Why can't the grid information be written to the file saved by the output writer (why does it have to be written to a separate file?)

Reply:
> Is this a constructor for a struct called WriteGeometry? I think it is not, in which case it should use snake_case.

It is a plain function. I'll convert to snake case in the next commit.

> In the rest of the code, I think we use the word grid to refer to the information that is saved by this function. We haven't introduced the concept of geometry --- I'm not sure this is the right place. What do you think?

We could call it `write_grid` instead of `write_geometry`. I used geometry because MOM6 called it that.  

> Why can't the grid information be written to the file saved by the output writer (why does it have to be written to a separate file?)

The outputwriter could be used to write global or sliced output. So if someone writes only sliced output, the grid in that file will also be sliced. In that case it will be useful to provide functionality to write the entire grid to an independent file.

Reply:
> We could call it write_grid instead of write_geometry. I used geometry because MOM6 called it that.

I feel it makes sense to use `grid` here since we use it elsewhere right now so that the name of the function is self-explanatory. If we think it makes sense to use `geometry` rather than `grid` throughout the code, we can perhaps discuss that in a future PR or issue?

Reply:
> The outputwriter could be used to write global or sliced output. So if someone writes only sliced output, the grid in that file will also be sliced.

Okay. Presumably this doesn't matter for grids with separable dimensions (like `RegularCartesianGrid`, but could matter for grids with curvature? 

I don't quite see why it helps to "slice" the grid for sliced output (is that just a feature to save memory)? Maybe you can explain why it's important to save subsets of the grid for certain purposes? Why not save the whole grid (and record the location of the sliced output within the grid)?

Reply:
> I feel it makes sense to use grid here since we use it elsewhere right now so that the name of the function is self-explanatory. If we think it makes sense to use geometry rather than grid throughout the code, we can perhaps discuss that in a future PR or issue?

No, I think grid is fine. It conveys what it is. I changed geometry to grid.

> Okay. Presumably this doesn't matter for grids with separable dimensions (like RegularCartesianGrid, but could matter for grids with curvature?

If the curved grid is a lat-lon grid, we can just store latC, lonC, latF, lonF, etc. 

A useful feature would be outputting dx and dy for all grids. So for implementing, say, a differential operator for postprocessing end users don't have to worry about the nature of the grid.

> I don't quite see why it helps to "slice" the grid for sliced output (is that just a feature to save memory)? Maybe you can explain why it's important to save subsets of the grid for certain purposes? Why not save the whole grid (and record the location of the sliced output within the grid)?

If you refer to my previous reply (where an ncdump is shown), a variable, say, `u` has dimensions `(Time,xF,yC,zC)`. Netcdf format imposes a condition that `size(u,1) == length(Time)`, `size(u,2) == length(zC)`, and so on. Tools like `xarray` leverage this property quite effectively. For example, slicing and plotting is a one liner in xarray:

```python
import xarray as xr
with xr.open_dataset('dump.nc') as ds
    ds.u.sel(Time=1,xF=10).plot()
````
That's it. This plots a yz section at x=10, Time=1. It figures out axes, labels them appropriately, etc in the background. 

If you don't slice the grid along with the varibale, dimension sizes won't align with the variable sizes and netcdf would just error out. You could just have integer indices as netcdf dimensions but then we lose all the benefits of the format. Does this help?

It would be nice to eventually have jld2 output writer do this too because then we might be able to write a jld2 to xarray interpreter. Xarray is, by far, the best tool right now to deal with multidimensional arrays with labeled axes. We could potentially attract more users if we start supporting it.

Reply:
Okay, I understand. I think I see how what you describe is a simple solution to the problem of describing subsets of data within a grid --- though in some ways this solution is necessary because NetCDF doesn't have an abstraction for these subsets of data... ? 

It would seem that the need to save the entire grid separately from subsets / slices of the grid within a particular NetCDF file is actually some kind of limitation, rather than a feature (presumably we would rather not have to do this). 

Either way, I agree it's good to have functionality that facilitates usage by popular packages. Since I don't completely understand the issue, if you say its necessary then I'm fine to keep this functionality to facilitate such a usage pattern.

As for facilitation of interop between `JLD2` and `xarray`, would it make sense to write a wholly separate output writer devoted to outputting fields and subsets of fields in `JLD2` format that can be read by `xarray`? As it is the `JLD2OutputWriter` is quite simple and therefore flexible (arbitrary data can be saved, including the result of complicated on-line calculations, arbitrary julia types, etc).

As an aside, it may also be interesting to take a look at

https://github.com/JuliaArrays/AxisArrays.jl

Reply:
> Okay, I understand. I think I see how what you describe is a simple solution to the problem of describing subsets of data within a grid --- though in some ways this solution is necessary because NetCDF doesn't have an abstraction for these subsets of data... ?

Exactly. Tools like xarray provide this abstraction on top of netcdf. I presume most users would want to store global datasets and then take slices using their tool of choice. This netcdfwriter stores global datasets if slices are not provided (well not quite, we need to discuss halos). Storing slices directly from the model makes sense only if someone is constrained by storage rather than computational resources.

> It would seem that the need to save the entire grid separately from subsets / slices of the grid within a particular NetCDF file is actually some kind of limitation, rather than a feature (presumably we would rather not have to do this).

Yes, but we write the global grid only once at the beginning of the run, so it's not expensive. Also, for sliced output, slices are computed when we initialize the outputwriter. So no new computation is carried out at every output iteration. We do use `getindex`, which creates copies of arrays, to get slices of fieldarrays at every output iteration. We could replace it with `view` if memory becomes a constraint.

Reply:
> As for facilitation of interop between JLD2 and xarray, would it make sense to write a wholly separate output writer devoted to outputting fields and subsets of fields in JLD2 format that can be read by xarray? As it is the JLD2OutputWriter is quite simple and therefore flexible (arbitrary data can be saved, including the result of complicated on-line calculations, arbitrary julia types, etc).

I don't understand the JLD2 file type well enough to make a call on this. @ali-ramadhan what do you think. 

Netcdf does have support for scalars and attributes. So may be we can store some arbitrary data in netcdf as well. Will look into it eventually.

Reply:
> Netcdf does have support for scalars and attributes. So may be we can store some arbitrary data in netcdf as well.

NetCDF cannot store arbitrary type information, however, which means that it cannot store julia types (like a `BoundaryCondition`). In `JLD2` a `BoundaryCondition` can be saved directly to disk and then loaded back without shenanigans or the need to manually reconstruct the julia type.

> Yes, but we write the global grid only once at the beginning of the run, so it's not expensive.

I'm referring to the annoyance of having to manually save the grid in a separate file via a manual call to to `write_grid`.

Reply:
> I'm referring to the annoyance of having to manually save the grid in a separate file by calling write_grid.

If the users write out a global output file it won't be necessary to call `write_grid`. 

> NetCDF cannot store arbitrary type information, however, which means that it cannot store julia types (like a BoundaryCondition). In JLD2 a BoundaryCondition can be saved directly to disk and then loaded back without shenanigans or the need to manually reconstruct the julia type.

I understand.